### PR TITLE
[ncp] split ncp_base and rearrange declarations.

### PR DESCRIPTION
**Pull request by bukepo**
This PR mainly splits ncp_base into smaller files.

* split handlers into general, mtd, ftd and radio-only.
* move ChangedPropsSet into separate file.
* add Ncp namespace.

### CODE REVIEW

**Comment by abtink** [comment]
@bukepo  thanks a lot for doing this. I was advocating/considering a breakdown of `ncp_base` file for a while. The breakdown you have sounds pretty good. FTD properties in one file, MTD ones in another and so on.

I have one big suggestion: Would it be possible to divide the changes into two commits/PRs. One PR  doing the splitting of the `ncp_base.cpp` code into multiple files...(with no new code/logic added)...  and then in second commit/PR to include all the new logic/changes for the raw mode. I think this will make it easier to track the changes and keeps the commits smaller and more manageable.... I know may be hard to split the changes, but would like to ask you to consider it. Thanks... :)

Thoughts?

**Comment by bukepo** [comment]
@abtink Thanks for your suggestions. I will update this PR.

File: src/ncp/ncp_base.cpp [hunk 1]
```diff
@@ -691,41 +562,42 @@ NcpBase::NcpBase(otInstance *aInstance):
 #if OPENTHREAD_ENABLE_RAW_LINK_API
     mCurTransmitTID(0),
     mCurReceiveChannel(OPENTHREAD_CONFIG_DEFAULT_CHANNEL),
-    mCurScanChannel(NCP_INVALID_SCAN_CHANNEL),
+    mCurScanChannel(kInvalidScanChannel),
```
Comment (abtink) [inline src/ncp/ncp_base.cpp]: 👍 for using new constant here...

File: src/ncp/ncp_base.hpp [hunk 2]
```diff
@@ -101,13 +106,75 @@ class NcpBase
...
+     *                        If aDataLen is non-zero, this param MUST NOT be NULL.
+     * @param[in]  aDataLen   The number of bytes of data from aDataPtr to send.
+     *
+     * @retval OT_ERROR_NONE         The data was queued for delivery to the host.
+     * @retval OT_ERROR_BUSY         There are not enough resources to complete this
+     *                               request. This is usually a temporary condition.
+     * @retval OT_ERROR_INVALID_ARGS The given aStreamId was invalid.
+     *
+     */
+    otError StreamWrite(int aStreamId, const uint8_t *aDataPtr, int aDataLen);
```
Comment (abtink) [inline src/ncp/ncp_base.hpp]: 👍 for adding documentation for `StreamWrite`....

File: src/ncp/ncp_base.hpp [hunk 3]
```diff
@@ -101,13 +106,75 @@ class NcpBase
...
+
+#if OPENTHREAD_MTD || OPENTHREAD_FTD
+#if OPENTHREAD_CONFIG_NCP_ENABLE_PEEK_POKE
```
Comment (abtink) [inline src/ncp/ncp_base.hpp]: Peek/poke support can be useful in general. How about not limiting them to `MTD` and `FTD`...?

File: src/ncp/ncp_base.cpp [hunk 4]
```diff
@@ -27,7 +27,7 @@
 
 /**
  * @file
- *   This file implements a Spinel interface to the OpenThread stack.
+ *   This file implements minimal thread device required Spinel interface to the OpenThread stack.
```
Comment (abtink) [inline src/ncp/ncp_base.cpp]: suggest changing "minimal thread device" to "general" here...

**Comment by abtink** [comment]
@jwhui can you please make sure when merging this the two commits are not squashed and go in seperately. Thanks.

**Comment by bukepo** [comment]
@abtink Thanks for your suggestions. I've updated comments and moved peek and poke into general part of `ncp_base`.

File: etc/visual-studio/libopenthread-ncp-spi.vcxproj [hunk 5]
```diff
@@ -58,18 +58,23 @@
     </ClCompile>
   </ItemDefinitionGroup>
   <ItemGroup>
+    <ClCompile Include="..\..\src\ncp\changed_props_set.cpp" />
     <ClCompile Include="..\..\src\ncp\ncp_base.cpp" />
+    <ClCompile Include="..\..\src\ncp\ncp_base_ftd.cpp" />
+    <ClCompile Include="..\..\src\ncp\ncp_base_mtd.cpp" />
+    <ClCompile Include="..\..\src\ncp\ncp_base_raw.cpp" />
```
Comment (jwhui) [inline etc/visual-studio/libopenthread-ncp-spi.vcxproj]: Using the name "raw" seems a bit underspecified.  How about simply "radio"?

File: src/ncp/changed_props_set.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+ *    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ *    ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ *    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include "changed_props_set.hpp"
+
+namespace ot {
```
Comment (jwhui) [inline src/ncp/changed_props_set.cpp]: should we add an `ncp` namespace?

File: src/core/api/link_raw.hpp [hunk 7]
```diff
@@ -53,69 +53,195 @@ class LinkRaw
...
+     */
+    uint16_t GetShortAddress(void) { return mShortAddress; }
+
+    /**
+     * This method updates short address.
+     *
+     * @param[in]   aShortAddress   The short address.
+     *
+     */
+    void SetShortAddress(uint16_t aShortAddress) { mShortAddress = aShortAddress; }
```
Comment (jwhui) [inline src/core/api/link_raw.hpp]: Should these setters call the `otPlatRadio*` APIs?

File: src/core/api/link_raw_api.cpp [hunk 8]
```diff
@@ -67,6 +84,9 @@ otError otLinkRawSetPanId(otInstance *aInstance, uint16_t aPanId)
     VerifyOrExit(aInstance->mLinkRaw.IsEnabled(), error = OT_ERROR_INVALID_STATE);
 
     otPlatRadioSetPanId(aInstance, aPanId);
```
Comment (jwhui) [inline src/core/api/link_raw_api.cpp]: Should we move calls to `otPlatRadio*` within `LinkRaw` methods?

File: src/core/api/link_raw.hpp [hunk 9]
```diff
@@ -53,69 +53,195 @@ class LinkRaw
...
 
+    /**
+     * This method is called when the transmission has started.
+     *
+     * @param[in]  aFrame     A pointer to the frame that is being transmitted.
+     *
+     */
+    void TransmitStarted(otRadioFrame *aFrame);
+
+#if OPENTHREAD_RAW
```
Comment (jwhui) [inline src/core/api/link_raw.hpp]: Why are we wrapping the methods below with `OPENTHREAD_RAW`?  If this class is included, shouldn't the functionality below also be available?

File: src/core/api/link_raw_api.cpp [hunk 10]
```diff
@@ -38,26 +38,43 @@
...
 otError otLinkRawSetEnable(otInstance *aInstance, bool aEnabled)
 {
     otError error = OT_ERROR_NONE;
 
-    VerifyOrExit(!aInstance->mThreadNetif.IsUp(), error = OT_ERROR_INVALID_STATE);
-
     otLogInfoPlat(aInstance, "LinkRaw Enabled=%d", aEnabled ? 1 : 0);
+#if OPENTHREAD_RAW
+
+    if (aEnabled)
```
Comment (jwhui) [inline src/core/api/link_raw_api.cpp]: Why not call `mLinkRaw.SetEnabled()` here?

File: src/core/api/link_raw_api.cpp [hunk 11]
```diff
@@ -259,11 +297,41 @@ otError otLinkRawSrcMatchClearExtEntries(otInstance *aInstance)
     return error;
 }
 
+#if OPENTHREAD_RAW
```
Comment (jwhui) [inline src/core/api/link_raw_api.cpp]: Do we need to wrap the functions below with `OPENTHREAD_RAW`?

File: src/core/api/link_raw.hpp (hunk 9 above)
Comment (bukepo) [inline src/core/api/link_raw.hpp]: If only raw link is enabled, `ncp_base` don't call these functions. Instead, it calls functions in the `link_api.cpp`, which calls some mac layer stuff. `link_api.cpp` is basically wrapper of mac layer, but `link_raw_api.cpp` is mostly wrapper of OpenThread's radio layer. However, maybe we could try merging `link_api.cpp` and `link_raw_api.cpp`. Thoughts?

File: src/core/api/link_raw_api.cpp (hunk 10 above)
Comment (bukepo) [inline src/core/api/link_raw_api.cpp]: In ncp's raw mode, link raw's functionality is always enabled. The `LinkRaw::mEnabled` is actually a flag of whether the link raw functionality is activated under the existing implementation, and it's not whether the radio layer is enabled or not. This PR doesn't change this design. Shall I change this in this PR?

File: src/core/api/link_raw_api.cpp (hunk 8 above)
Comment (bukepo) [inline src/core/api/link_raw_api.cpp]: I didn't put them within `LinkRaw` because those functions are not always necessary under the current design. `ncp_base` would sometimes call versions in `link_api.cpp`.

Comment (jwhui) [inline src/core/api/link_raw_api.cpp]: Yes, I see the discrepancy now.  The vision was always to have a common API for operating in "radio-only mode" regardless of whether or not OpenThread core was compiled in or out.  Do you think we can work towards that vision?

Comment (bukepo) [inline src/core/api/link_raw_api.cpp]: Thanks, I will update this PR.

Comment (bukepo) [inline src/core/api/link_raw_api.cpp]: In order not to block other PR, I think it will be better to submit another PR for radio-only task. This PR only splits `ncp_base` and adding namespace `ot::Ncp`.

Comment (jwhui) [inline src/core/api/link_raw_api.cpp]: Sounds good.
